pkg/controllers/machinesync/machine_sync_controller.go

```diff
@@ -100,6 +112,73 @@ func (r *MachineSyncReconciler) SetupWithManager(mgr ctrl.Manager) error {
 
 // Reconcile reconciles CAPI and MAPI machines for their respective namespaces.
 func (r *MachineSyncReconciler) Reconcile(ctx context.Context, req reconcile.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx, "namespace", req.Namespace, "name", req.Name)
+
+	logger.V(1).Info("Reconciling machine")
+	defer logger.V(1).Info("Finished reconciling machine")
+
+	var (
+		mapiMachineNotFound bool
+		capiMachineNotFound bool
+	)
+
+	// Get the MAPI Machine
+	mapiMachine := &machinev1beta1.Machine{}
+	if err := r.Get(ctx, req.NamespacedName, mapiMachine); apierrors.IsNotFound(err) {
+		logger.Info("MAPI Machine not found")
+
+		mapiMachineNotFound = true
+	} else if err != nil {
+		logger.Error(err, "Failed to get MAPI Machine")
+		return ctrl.Result{}, fmt.Errorf("%w: %w", errFailedToGetMAPIMachine, err)
+	}
+
+	// Get the corresponding CAPI Machine
+	capiMachine := &capiv1beta1.Machine{}
+	capiNamespacedName := client.ObjectKey{
+		Namespace: r.CAPINamespace,
+		Name:      req.Name,
+	}
+
+	if err := r.Get(ctx, capiNamespacedName, capiMachine); apierrors.IsNotFound(err) {
+		logger.Info("CAPI Machine not found")
+
+		capiMachineNotFound = true
+	} else if err != nil {
+		logger.Error(err, "Failed to get CAPI Machine")
+		return ctrl.Result{}, fmt.Errorf("%w: %w", errFailedToGetCAPIMachine, err)
+	}
+
+	if mapiMachineNotFound && capiMachineNotFound {
+		logger.Info("CAPI and MAPI machines not found, nothing to do")
+		return ctrl.Result{}, nil
+	}
+
+	// We mirror if the CAPI machine is owned by a MachineSet which has a MAPI
+	// counterpart.
+	if mapiMachineNotFound {
+		if shouldReconcile, err := r.shouldMirrorCAPIMachine(ctx, capiMachine, logger); err != nil {
+			return ctrl.Result{}, err
+		} else if shouldReconcile {
+			return r.reconcileCAPItoMAPI(ctx, capiMachine, mapiMachine)
+		}
+	}
+
+	// If we have not set AuthoritativeAPI, we default to MachineAuthorityMachineAPI
+	if mapiMachine.Spec.AuthoritativeAPI == machinev1beta1.MachineAuthorityClusterAPI {
+		return r.reconcileCAPItoMAPI(ctx, capiMachine, mapiMachine)
+	} else {
+		return r.reconcileMAPItoCAPI(ctx, mapiMachine, capiMachine)
+	}
+}
+
+// reconcileCAPItoMAPI reconciles a CAPI Machine to a MAPI Machine.
+func (r *MachineSyncReconciler) reconcileCAPItoMAPI(ctx context.Context, capiMachine *capiv1beta1.Machine, mapiMachine *machinev1beta1.Machine) (ctrl.Result, error) {
```

Comment:
I'd also make these more descriptive to avoid mixup :)
```suggestion
func (r *MachineSyncReconciler) reconcileCAPIMachinetoMAPIMachine(ctx context.Context, capiMachine *capiv1beta1.Machine, mapiMachine *machinev1beta1.Machine) (ctrl.Result, error) {
```